### ISSUE TITLE
Error: Cannot find module '../data/config/data.json'

### ISSUE THREAD

**Issue by nouryf** [tobinbradley/quality-of-life-dashboard#9]
**Environment**: Using WSL Ubuntu 20, npm 6.14.4 and nodejs v10.19.0, I tried to build and run the project but had this error:

```
 qol-dashboard@1.0.0 prebuild /mnt/c/Dev/quality-of-life-dashboard
> node build/datagen.js

internal/modules/cjs/loader.js:638
    throw err;
    ^

Error: Cannot find module '../data/config/data.json'
    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:636:15)
    at Function.Module._load (internal/modules/cjs/loader.js:562:25)
    at Module.require (internal/modules/cjs/loader.js:692:17)
    at require (internal/modules/cjs/helpers.js:25:18)
    at Object.<anonymous> (/mnt/c/Dev/quality-of-life-dashboard/build/datagen.js:3:18)
    at Module._compile (internal/modules/cjs/loader.js:778:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:789:10)
    at Module.load (internal/modules/cjs/loader.js:653:32)
    at tryModuleLoad (internal/modules/cjs/loader.js:593:12)
    at Function.Module._load (internal/modules/cjs/loader.js:585:3)
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! qol-dashboard@1.0.0 prebuild: `node build/datagen.js`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the qol-dashboard@1.0.0 prebuild script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in: ....
```

**Here are the followed steps:**

- Cloned the repo
- Run `npm install` (got some warnings about deprecated versions but no errors)
- Clone the data repo to `data` under "the quality of life dashboard" folder 
- In the "data" sub-folder I run `npm install`: warnings but no errors.
- Run `npm run build`: data were downloaded and no errors
- Went back to "the quality of life dashboard" folder and run `npm start`
- I got the error listed above: Cannot find module '../data/config/data.json'

I'm doing some thing wrong or this is a WSL related issue ?

Thanks

**Comment by nouryf** [comment]
I just run the same steps on a droplet (Digital Ocean) and got the same error. The OS was "Ubuntu 20.04.4 LTS". So, it does not look like a WSL issue.

**Comment by tobinbradley** [comment]
Not sure what's going on there. I assume there is a data/config/data.json file? I just tried running it on a fresh clone and can't get even get that far. I'm on Node 17 and apparently a dependency is using a deprecated Node something or another. 

I hate to ask you to try cloning something else, but it'll be faster than trying to fix this repo. I recently (last week) finished refactoring the code base in [another repo](https://github.com/tobinbradley/quality-of-life-explorer). It's a huge improvement. I just tried a cold clone to make sure I'm not setting you up for more disappointment and everything is working.

```bash
git clone https://github.com/tobinbradley/quality-of-life-explorer.git dashboard
cd dashboard
npm install
npm run prebuild
npm run dev
```

I'll stick a "stuff has moved" header at the top of the README here. Apologies, and thank you for reporting the problem!

**Comment by nouryf** [comment]
Thank you for the quick response. I just tried the new repo and it did error at the last step (`npm run dev`):

```
npm run dev

> quality-of-life-explorer@4.0.0 dev /mnt/c/Dev/dashboard
> vite

internal/modules/cjs/loader.js:638
    throw err;
    ^

Error: Cannot find module 'worker_threads'
    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:636:15)
    at Function.Module._load (internal/modules/cjs/loader.js:562:25)
    at Module.require (internal/modules/cjs/loader.js:692:17)
    at require (internal/modules/cjs/helpers.js:25:18)
    at Object.<anonymous> (/mnt/c/Dev/dashboard/node_modules/vite/dist/node/chunks/dep-9c153816.js:25:20)
    at Module._compile (internal/modules/cjs/loader.js:778:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:789:10)
    at Module.load (internal/modules/cjs/loader.js:653:32)
    at tryModuleLoad (internal/modules/cjs/loader.js:593:12)
    at Function.Module._load (internal/modules/cjs/loader.js:585:3)
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! quality-of-life-explorer@4.0.0 dev: `vite`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the quality-of-life-explorer@4.0.0 dev script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in: ...
```
I can send the log if it may help shed some light on this.

**Comment by nouryf** [comment]
Oups, I did not answer your question (I assume there is a data/config/data.json file? ). I'll redo the steps as I deleted the local repo and see if that file is there. I don't think it was as it was the first thing I check but let me double check.

**Comment by nouryf** [comment]
`worker_threads` was apparently not supported in my Node version (v10.19.0) according to [this](https://stackoverflow.com/questions/62280966/cannot-find-module-worker-threads-error-in-production). I'll upgrade my Node and try again.

**Comment by tobinbradley** [comment]
No worries about checking that. I was just curious.

I think the other error is a problem with your Node version. worker_threads came to Node in v11. The current stable release of Node is 16.x, so you're a couple of stable releases out of date. I don't think there's a workaround for that other than updating Node.

**Comment by nouryf** [comment]
Yes that was indeed the issue. I upgraded my Node to version 16 and was able to rebuild and run the app just fine. Thanks a bunch for your help.

Can I override the used data by pointing to a different county through a FIPS change in [quality-of-life-data-uscensus](https://github.com/tobinbradley/quality-of-life-data-uscensus)  after cloning it in "data" sub-folder under the "dashboard" folder ? Or [quality-of-life-explorer](https://github.com/tobinbradley/quality-of-life-explorer) project does not support the data override mechanism ?

Thanks

**Comment by tobinbradley** [comment]
Well, sorta, but as the data folder config has changed there would be a lot of hand waving. Let me see if I can patch up that census repo so it'll work with the new repo. The new data folder is a lot simpler but it is definitely different.

**Comment by tobinbradley** [comment]
OK, I fixed up the census data thing. You should be able to give it a go. Just follow the directions there carefully. You'll have to manually move a couple of files. Also, I tweaked a few things from the explorer to accommodate different geography, so you'll want to git pull those changes.

**Comment by nouryf** [comment]
Great. Thank you. 
I did give it a try after changing the FIPS code to 36083 (Rensselaer County, NY), but it did not display the updated map nor the new data. I did the steps as outlined in the data [repo](https://github.com/tobinbradley/quality-of-life-data-uscensus#building-data). The .csv data were indeed fetched. I noticed the following:
- Under `metric` folder, the fetched data files were named `r1.csv, ..., r18.csv` (may be the `r` is for rensselaer) but under `meta` folder the .md documents were named `m1.md, ..., m18.md` .
- In `data.json` the metrics were named "m1", "m22". The count in that json is different from the count of the .md and the .csv files
- There is still a `positron-mecklenburg.json` that defines the bounds, but may be that is not used and instead the copied that `gl-style.json` that is indeed used. This one does not have the bounds, however.

When I said, the data were not displayed nor the correct map, here is what I saw:
![image](https://user-images.githubusercontent.com/18246160/160081177-02ec39ff-8963-466a-975a-72fc05fca6a5.png)

**Comment by tobinbradley** [comment]
Since I'm seeing the Mecklenburg map style I think you missed some steps. You'll see some pretty/ugly stamen tiles if you get it right. There was a file naming problem on some variables (I had to remove a few as the Census API was returning nothing), but it's patched now.

Try doing all this verbatim.

```bash
git clone https://github.com/tobinbradley/quality-of-life-explorer.git qolcensus
cd qolcensus
npm install
rm -rf data
git clone https://github.com/tobinbradley/quality-of-life-data-uscensus.git data
cd data
npm install
npm run build
cd ..
cp data/geojson/geography.geojson.json public/data/geography/geography.geojson.json
cp data/gl-style.json src/assets/gl-style.json
npm run prebuild
npm run dev
```

Here's our neighboring County:

![snapshot](https://i.imgur.com/SUhIg7c.png)

**Comment by nouryf** [comment]
Not sure what I did wrong but the steps your outlined are the one I used before. Any way, I did try them now, cloned to new folders and built every thing after changing the FIPS code. It did work great:

![image](https://user-images.githubusercontent.com/18246160/160159346-d2cbc4bb-4f7d-4f99-8dae-0045e8164ec5.png)

Few things need manual updates of course and this is expected like using vector tiles for the new county, search, title, jurisdiction, city council, etc. These are all fines and can be updated to reflect the new county.

One thing however that I'm not sure about is one of the metric "White or Caucasian" the numbers are a bit high for percentages. They should be <= 100%. I looked at the fetched .csv and they were indeed the number showed in the map. May be a discrepancy is some where else.

Much thanks for your help
